Firestore/core/src/util/byte_stream_apple.mm

```diff
@@ -64,24 +61,31 @@
   return read_result;
 }
 
-StreamReadResult ByteStreamNSInputStream::Read(size_t max_length) {
-  auto read = ReadToBuffer(max_length);
+StreamReadResult ByteStreamApple::Read(size_t max_length) {
+  // Serve from buffer_
+  if (buffer_.size() >= max_length) {
+    std::string result = buffer_.substr(0, max_length);
+    buffer_.erase(0, max_length);
+    auto read_result = StreamReadResult(std::move(result), eof());
+    return read_result;
+  }
+
+  auto read = ReadToBuffer(max_length - buffer_.size());
   if (read < 0) {
     return ErrorResult();
   }
   if (read == 0) {
     return EofResult();
   }
 
-  std::string buffer_copy(buffer_);
+  auto read_result = StreamReadResult(std::move(buffer_), eof());
```

Comment:
Now that this pattern is repeated a few times, what do you think about creating a helper function?
```cpp
StreamReadResult ConsumeBuffer(bool is_eof) {
  StreamReadResult result(std::move(buffer_), is_eof);
  buffer_.clear();
  return result;
}

// ...
return ConsumeBuffer(eof());
```